### ISSUE TITLE
Improve compatibility messaging for users.

### ISSUE THREAD

**Issue by tchaton** [Lightning-AI/lightning#13823]
### Motivation

After reading through PyTorch codebase, I came across this code example:

```
from typing import Any, Dict
import textwrap

_BACK_COMPAT_OBJECTS : Dict[Any, None] = {}
_MARKED_WITH_COMATIBLITY : Dict[Any, None] = {}

def compatibility(is_backward_compatible : bool):
    if is_backward_compatible:

        def mark_back_compat(fn):
            docstring = textwrap.dedent(getattr(fn, '__doc__', None) or '')
            docstring += """
.. note::
    Backwards-compatibility for this API is guaranteed.
"""
            fn.__doc__ = docstring
            _BACK_COMPAT_OBJECTS.setdefault(fn)
            _MARKED_WITH_COMATIBLITY.setdefault(fn)
            return fn

        return mark_back_compat
    else:

        def mark_not_back_compat(fn):
            docstring = textwrap.dedent(getattr(fn, '__doc__', None) or '')
            docstring += """
.. warning::
    This API is experimental and is *NOT* backward-compatible.
"""
            fn.__doc__ = docstring
            _MARKED_WITH_COMATIBLITY.setdefault(fn)
            return fn

        return mark_not_back_compat
```

I believe such logic would provide value.

Furthermore, I believe we could create a mechanism to better inform users about depreciation. 

Here is a proposed mechanism:

# On those calls

trainer = Trainer(...)
trainer.X(...)

# Find all the non-built classes and activate a mechanism so we can capture all depreceated calls from external classes to our own codebase.

Example:

# pytorch_lightning/strategies/dpp.py

class DDPStrategy():

    @depreceated("version_")
    def method_X(self, ...)
        ...

# users/..../custom_strategies/custom_dpp.py

class CustomDDPStrategy(CustomDDPStrategy):

    def method_X(self, ...):
        super().method_X(...)

trainer = Trainer(strategy=CustomDDPStrategy(...))

```
Warning Reports:

`CustomDDPStrategy.method_X` calls `DDPStrategy.method_X` which is depreceated in 1.X and remove in 1.Y. Use method_Z instead in `CustomDDPStrategy`.
...
...
```` 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

**Comment by tchaton** [comment]
@carmocca for thoughts.

**Comment by carmocca** [comment]
I don't see the advantage over our current system of showing deprecation messages.

**Comment by tchaton** [comment]
@carmocca I believe the trace is helpful and users are usually complaining about the long lists of warnings without a clear explanation about the root cause.

**Comment by carmocca** [comment]
As discussed online, I would give the option to include the stacktrace at the time of a deprecation.

This could be done externally with this addition:


```python
import traceback
import warnings
import sys


def showwarning_with_deprecation_traceback(message, category, filename, lineno, file=None, line=None):
    log = file if hasattr(file, 'write') else sys.stderr
    if issubclass(category, DeprecationWarning):
        stack = traceback.extract_stack()[:-5]  # `rank_zero_deprecation` will add 5 extra levels
        traceback.print_list(stack, file=log)
    log.write(warnings.formatwarning(message, category, filename, lineno, line))


warnings.showwarning = showwarning_with_deprecation_traceback
```

example:

```python
  rank_zero_warn(
  File "/home/carmocca/git/lightning/examples/pl_bug_report/bug_report_model.py", line 82, in <module>
    run()
  File "/home/carmocca/git/lightning/examples/pl_bug_report/bug_report_model.py", line 61, in run
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 700, in fit
    self._call_and_handle_interrupt(
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 654, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 741, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 1166, in _run
    results = self._run_stage()
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 1252, in _run_stage
    return self._run_train()
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 1282, in _run_train
    self.fit_loop.run()
  File "/home/carmocca/git/lightning/src/pytorch_lightning/loops/loop.py", line 195, in run
    self.on_run_start(*args, **kwargs)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/loops/fit_loop.py", line 210, in on_run_start
    self.trainer.reset_train_dataloader(self.trainer.lightning_module)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 1832, in reset_train_dataloader
    apply_to_collection(loaders, DataLoader, self._data_connector._worker_check, "train_dataloader")
  File "/home/carmocca/git/lightning/src/pytorch_lightning/utilities/apply_func.py", line 100, in apply_to_collection
    return function(data, *args, **kwargs)
  File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/connectors/data_connector.py", line 226, in _worker_check
    rank_zero_deprecation("foobar")
/home/carmocca/git/lightning/src/pytorch_lightning/trainer/connectors/data_connector.py:226: LightningDeprecationWarning: foobar
  rank_zero_deprecation("foobar")
```

**Comment by deependujha** [comment]
Closing this out since it's been inactive and didn't reach a definitive implementation plan beyond the existing deprecation system.
